### ISSUE TITLE
`bad pattern: funcs=(\n      ~zsh-autocomplete/functions/.autocomplete.*~*.zwc(N-.:a)\n  )`

### ISSUE THREAD

**Issue by muuvmuuv** [marlonrichert/zsh-autocomplete#335]
* `zsh-autocomplete` version: latest I assume
* Zsh version: zsh 5.8 (x86_64-apple-darwin20.0)
* Framework: Starship
* Plugin manager: zinit
* Operating system: macOS ARM latest release

I was not able to reproduce it with the template you provided because I only get this error on loading the shell:

```
~/Development
;
(anon):22: bad pattern: funcs=(\n      ~zsh-autocomplete/functions/.autocomplete.*~*.zwc(N-.:a)\n  )
```

My zshrc:

```zsh
eval "$(starship init zsh)"

export VOLTA_HOME="$HOME/.volta"
export PATH="$VOLTA_HOME/bin:$PATH"

alias vim="nvim"
alias dir="ls -lah"
alias zshrc="code ~/.zshrc"

# Globbing

setopt NO_CASE_GLOB
setopt EXTENDED_GLOB

# History

HISTORY_IGNORE="(ls|dir|cd|pwd|exit|cd ..|..|*venv*)"

setopt EXTENDED_HISTORY
setopt INC_APPEND_HISTORY
setopt HIST_EXPIRE_DUPS_FIRST
setopt HIST_REDUCE_BLANKS
setopt HIST_IGNORE_ALL_DUPS
setopt HIST_IGNORE_SPACE
setopt HIST_IGNORE_DUPS
setopt HIST_FIND_NO_DUPS
setopt HIST_SAVE_NO_DUPS

# Misc

setopt AUTO_CD
setopt MENU_COMPLETE
setopt NO_LIST_AMBIGUOUS

# Settings

zstyle ':autocomplete:*' min-input 2
zstyle ':autocomplete:tab:*' fzf-completion yes

# Plugins

source "$HOME/.zinit/bin/zinit.zsh"
autoload -Uz _zinit
(( ${+_comps} )) && _comps[zinit]=_zinit

zinit wait lucid light-mode for \
  marlonrichert/zsh-autocomplete \
  atinit"zicompinit; zicdreplay" \
      zdharma/fast-syntax-highlighting \
  blockf atpull'zinit creinstall -q .' \
      zsh-users/zsh-completions

zinit ice from"gh-r" as"program" bpick"*arm*"
zinit light junegunn/fzf
```

My zsh_history looks like this:

```
: 1631085612:0;cd app
```

**Comment by muuvmuuv** [comment]
Just tried with the Homebrew zsh Version "zsh 5.8 (arm-apple-darwin20.2.0)" same error.

**Comment by marlonrichert** [comment]
It's actually a bug in Zinit: https://github.com/zdharma/zinit/issues/366

I had a workaround for it, but it looks like it doesn't work anymore. I will add a new workaround.

However, I do strongly recommend you switch to a plugin manager that doesn't [do as much nonsense](https://github.com/zdharma/zinit/blob/f3cd0b78950478207a9a5308532781a85c5baacb/zinit.zsh#L675).

**Comment by muuvmuuv** [comment]
I am open for suggestions. I only have fzf and two other zsh plugins I need. I would even add them in a fast way natively without a pm but I have no clue how to do that. I will take a closer look at that the next days, maybe I really dont need any.

**Comment by marlonrichert** [comment]
If you do want to use a plugin manager, I would naturally recommend [Znap](https://github.com/marlonrichert/zsh-snap). 🙂

**Comment by muuvmuuv** [comment]
Okay wow you are amazing. Now I don't even see any blink or delay its just up in 0.1ms xD Combined it with Starship.rs and it's just the perfect fit for it.

**Comment by marlonrichert** [comment]
Great to hear! 😃 

---

PS: If you enjoy using my software, please consider [sponsoring me](https://github.com/sponsors/marlonrichert).

**Comment by weilbith** [comment]
This issue is still there, right?

**Comment by muuvmuuv** [comment]
No, gone after updating, at least on my end.

**Comment by marlonrichert** [comment]
@weilbith Can you give me a reproducible test case?

**Comment by weilbith** [comment]
Hmm. I just created a new user. And then I do:

```zsh
$ zsh
% source /usr/share/zinit/zinit.zsh
% zinit lucid light-mode for marlonrichert/zsh-autocomplete
# installation output...
(anon):22: bad pattern: funcs=(\n     ~zsh-autocomplete/functions/.autocomplete.*~*.zwc(N-.:a)\n  )
%
```

That appears on each new shell I open. Autocomplete doesn't work at all (god damn I love your plugin so much it is a nightmare without). And it is a little worse because Powerlevel10k instant prompt reports tons of output because something is logging before the first prompt. Very annoying atm.

**Comment by jack1142** [comment]
I'm seeing the same issue with [zcomet](https://github.com/agkozak/zcomet):
```
(anon):22: bad pattern: funcs=(\n      ~zsh-autocomplete/functions/.autocomplete.*~*.zwc(N-.:a)\n  )
Cannot source /home/ubuntu/.zcomet/repos/marlonrichert/zsh-autocomplete/zsh-autocomplete.plugin.zsh.
ubuntu-virtual-machine% 
```

Contents of `.zshrc`:
```
# Clone zcomet if necessary
if [[ ! -f ${ZDOTDIR:-${HOME}}/.zcomet/bin/zcomet.zsh ]]; then
  command git clone https://github.com/agkozak/zcomet.git ${ZDOTDIR:-${HOME}}/.zcomet/bin
fi

source ${ZDOTDIR:-${HOME}}/.zcomet/bin/zcomet.zsh

zcomet load marlonrichert/zsh-autocomplete
```

**Comment by marlonrichert** [comment]
I didn’t realize it before, but this issue is a duplicate of #339. Also, 96fec14c2963d06412af82c9873c5ff64336a1a3 doesn’t fix the problem you have; it’s  a workaround  for another Zinit problem. Apologies for the confusion.

@weilbith It’s caused by a bug in Z Init: https://github.com/zdharma/zinit/issues/543. It’s not that hard to fix, though. See [my comment there](https://github.com/zdharma/zinit/issues/543#issuecomment-920677219) for an explanation.

@jack1142 Looks like Z Comet has the same bug as Z Init. Again, see [my comment on the Z Init issue](https://github.com/zdharma/zinit/issues/543#issuecomment-920677219) for an explanation. When you open an issue on Z Comet’s repo, make sure you include a link to my comment.

And in case you didn’t know it: I maintain a fast, lightweight & simple plugin manager called [Znap](https://github.com/marlonrichert/zsh-snap), which naturally doesn’t suffer from the same problem. 😉

**Comment by marlonrichert** [comment]
> No, gone after updating, at least on my end.

@muuvmuuv Because you are using Znap now, right? Because the problem still hasn’t been fixed in Z Init. 🙂

**Comment by weilbith** [comment]
> And in case you didn’t know it: I maintain a fast, lightweight & simple plugin manager called Znap, which naturally doesn’t suffer from the same problem. wink

I know. I'm thinking about alternatives. But Znap still has no lazy loading, correct?

> @weilbith It’s caused by a bug in Z Init: zdharma/zinit#543. It’s not that hard to fix, though. See my comment there for an explanation.

So I wen to `$XDG_CACHE_HOME/zinit/plugins/marlonrichert---zsh_autocomplete` and run `zcompile -R zsh-autocomplete.plugin..zsh` there. No autocompletion works but I get tons of errors on a new shell which looks like that:

```
compinit:503: no such file or directory: /home/thore/.cache/zinit/completions/_yaourt
compinit:shift:505: shift count must be <= $#
compinit:503: no such file or directory: /home/thore/.cache/zinit/completions/_yarn
compinit:shift:505: shift count must be <= $#
compinit:503: no such file or directory: /home/thore/.cache/zinit/completions/_zcash-cli
compinit:shift:505: shift count must be <= $#
```

**Comment by marlonrichert** [comment]
> But Znap still has no lazy loading, correct?

No, not correct. 🙂  Znap _does_ have lazy loading:
```
% znap help function
Usage: znap function <func name> ... <init code>
Create a set of lazy-loading functions.
```

For example, to lazy-load `pyenv init` and `pyenv` completion:
```zsh
znap function pyenv _pyenv 'eval "$( pyenv init - --no-rehash )"'
compctl -K _pyenv pyenv
```
This way, `pyenv init` will be executed only once you run `pyenv` or try to invoke completion on it. 🙂

**Comment by marlonrichert** [comment]
> So I wen to `$XDG_CACHE_HOME/zinit/plugins/marlonrichert---zsh_autocomplete` and run `zcompile -R zsh-autocomplete.plugin..zsh` there. No autocompletion works but I get tons of errors on a new shell which looks like that:
> 
> ```
> compinit:503: no such file or directory: /home/thore/.cache/zinit/completions/_yaourt
> compinit:shift:505: shift count must be <= $#
> compinit:503: no such file or directory: /home/thore/.cache/zinit/completions/_yarn
> compinit:shift:505: shift count must be <= $#
> compinit:503: no such file or directory: /home/thore/.cache/zinit/completions/_zcash-cli
> compinit:shift:505: shift count must be <= $#
> ```

I cannot reproduce that. Here's what I did:
```zsh
% cd $(mktemp -d)
git clone --depth 1 -- https://github.com/zdharma/zinit.git $PWD/.zinit/bin
> .zshrc <<EOF
PS1="%# " PS2="  " RPS2="< %^"; setopt transientrprompt
source $PWD/.zinit/bin/zinit.zsh
zinit load marlonrichert/zsh-autocomplete
EOF
env -i HOME=$PWD ZDOTDIR=$PWD FPATH=$FPATH TERM=$TERM zsh -d

/tmp/tmp.5y9lh2d3MO/
Cloning into '/tmp/tmp.5y9lh2d3MO/.zinit/bin'...
remote: Enumerating objects: 180, done.
remote: Counting objects: 100% (180/180), done.
remote: Compressing objects: 100% (163/163), done.
remote: Total 180 (delta 10), reused 120 (delta 8), pack-reused 0
Receiving objects: 100% (180/180), 1.04 MiB | 1.34 MiB/s, done.
Resolving deltas: 100% (10/10), done.

Downloading marlonrichert/zsh-autocomplete...

Cloning into '/tmp/tmp.5y9lh2d3MO/.zinit/plugins/marlonrichert---zsh-autocomplete'...
▗ █████████████████████████ OBJ: 100, PACK: 3247/3247, COMPR: 100%, REC: 100%, RES: 100%          
Note: Compiling: zsh-autocomplete.plugin.zsh... OK.
Installed 10 completions. They are stored in $INSTALLED_COMPS array.
(anon):22: bad pattern: funcs=(\n      ~zsh-autocomplete/functions/.autocomplete.*~*.zwc(N-.:a)\n  )
% zcompile -R .zinit/plugins/marlonrichert---zsh-autocomplete/zsh-autocomplete.plugin.zsh
% exec zsh -d
%
```
And now autocompletion works just fine.

You have probably set some shell options that change the way `zcompile` parses the code. Try this instead:
```zsh
% emulate zsh -c 'zcompile -R .zinit/plugins/marlonrichert---zsh-autocomplete/zsh-autocomplete.plugin.zsh'
``` 

This will invoke `zcompile` with default shell options, which is what Z Init and Z Comet should be doing, too.

**Comment by muuvmuuv** [comment]
@marlonrichert Oh, I think you are right ^^ Did not think about that to be an issue with the pm rather than the plugin itself.

**Comment by marlonrichert** [comment]
Please disregard my comments so far as to what is the cause of the problem: On further investigation, it turns out that, actually, the only thing that fixes the problem, for both Z Init and Z Comet, is to `zcompile` from the command line. Changing the `zcompile` statement inside the code does nothing to solve it. There must be _something_ that both of them change about the compilation progress, but I don't know what it is. 😕

**Comment by muuvmuuv** [comment]
At all here… Just use [**znap**](https://github.com/marlonrichert/zsh-snap) xD I really have less problems with it (thanks @marlonrichert). Can strongly recommend it!

**Comment by agkozak** [comment]
I can run this at the command line:

    zsh -f
    () { zcompile -R zsh-autocomplete.plugin.zsh; }
    source zsh-autocomplete.plugin.zsh

with the same results:

    (anon):22: bad pattern: funcs=(\n      ~zsh-autocomplete/functions/.autocomplete.*~*.zwc(N-.:a)\n  )

**Comment by marlonrichert** [comment]
I finally managed to figure it out: It's an unfortunate way in which compiled or autoloaded code in Zsh is not 100% equivalent to the original source code. But now that I now what causes it, I also know how to work around it.

@agkozak Thanks for your help in tracking down the root cause of the problem!

**Comment by agkozak** [comment]
I suspect the reason Znap didn't have a problem was because it already has `zsh/param/private` loaded when `zcompile` is run. I tested that out of the command line.

**Comment by weilbith** [comment]
I can confirm that my `HEAD` is 8877a3b76c030f446ffe0c35a0acd0f9cf627081 but I still have this issue. 🤷🏾

**Comment by marlonrichert** [comment]
Do `git pull` again. It should be fixed now.

**Comment by weilbith** [comment]
> Delete what?

For some reason I had to delete the local clone of the plugin my plugin manager did. In this exact example for my I had to delete `$XDG_CACHE_HOME/zinit/plugins/marlonrichert---zsh-autocomplete`. It then reinstalled it and then it was fine. The `HEAD` of the commit before I deleted the directory and afterwards was the same. Dunno what was the difference here. Maybe some "compile" hook or so? 🤷🏾

**Comment by marlonrichert** [comment]
@agkozak I noticed that Z Comet is actually not that different in syntax from Znap and that you even copied some features from me. Rather than compete with each other, would you be interested in joining forces and merge our plugin managers into one?
